### PR TITLE
Added Elixir Loqui client

### PR DESCRIPTION
**Pull request by scohen**
Added an elixir client and did some slight refactoring to the server
as well. The new client's parsing takes advantage of erlang's binary
optimizations to not split incoming binaries until all the packets are
parsed. This should have some impact on the server, so I changed it
there too.

### CODE REVIEW

**Comment by scohen** [comment]
@fishcakez, can you take a look.

File: ex/loqui/lib/loqui/protocol/parser.ex [hunk 1]
```diff
@@ -0,0 +1,77 @@
+defmodule Loqui.Protocol.Parser do
+  use Loqui.{Opcodes, Types}
+
+  def parse(<<>>, extra_data),
+    do: parse(extra_data)
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/protocol/parser.ex]: nice optimisation to avoid naive copying `extra_data` most times.

File: ex/loqui/lib/loqui/protocol/parser.ex [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  defp do_parse_packet(<<@opcode_goaway, flags::uint8, close_code::uint16, payload_size::uint32, payload_data::binary-size(payload_size), rest::binary>>, accum) do
+    do_parse_packet(rest, [{:go_away, flags, close_code, payload_data} | accum])
+  end
+
+  defp do_parse_packet(<<@opcode_error, flags::uint8, sequence::uint32, error_code::uint16, payload_size::uint32, payload_data::binary-size(payload_size), rest::binary>>, accum) do
+    do_parse_packet(rest, [{:go_away, flags, sequence, error_code, payload_data} | accum])
+  end
+
+  defp do_parse_packet(leftover_data, []) do
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/protocol/parser.ex]: Binary optimised :+1:

File: ex/loqui/lib/loqui/protocol/frames.ex [hunk 3]
```diff
@@ -1,6 +1,10 @@
-defmodule Loqui.Frames do
+defmodule Loqui.Protocol.Frames do
   use Loqui.{Opcodes, Types}
 
+  def hello(flags, payload_data) do
+    <<@opcode_hello, flags :: uint8, 1::uint8, byte_size(payload_data) :: uint32, payload_data :: binary>>
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/protocol/frames.ex]: use iodata when writing for less copying (nb could be `IO.iodata_length(payload_data)` if iodata support.)
```elixir
[@opcode_hello, flags,<<byte_size(payload_data) :: unit32>> | payload_data]
```

File: ex/loqui/lib/loqui/client.ex [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+      codec: Codecs.Erlpack,
+      waiters: %{}
+  end
+
+  alias Loqui.Protocol.{Codec, Codecs, Compressor, Compressors, Frames, Parser}
+  use Connection
+  use Loqui.{Opcodes, Types}
+  require Logger
+
+  @type sequence :: pos_integer
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: This could be more accurately expressed as `1..unquote(@max_sequence)` if attribute registered before the type

File: ex/loqui/lib/loqui/client.ex [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    end
+  end
+
+  def disconnect(info, %State{sock: sock}=state) do
+    :ok = :gen_tcp.close(sock)
+    case info do
+
+      {:close, from} ->
+        Connection.reply(from, :ok)
+        {:stop, :normal, nil}
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: `nil` state?

File: ex/loqui/lib/loqui/client.ex [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+
+      {:noreply, %State{state | buffer: leftover_data}}
+    else
+
+      {:error, {:need_more_data, data}} ->
+        {:noreply, %State{buffer: data }}
+    end
+  end
+
+  def handle_info({:tcp_closed, _socket}, _state) do
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: OS could send a error and get `{:tcp_error, _, _}` too. e.g. if someone used `tcp_opts: [show_econnreset: true]`

File: ex/loqui/lib/loqui/client.ex [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  defp handle_packet({:ping, _flags, seq}, %State{sock: sock}=state) do
+    :gen_tcp.send(sock, Frames.pong(0, seq))
+    state
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{waiters: waiters}=state) do
+    me = self()
+    new_waiters =
+      case Map.pop(waiters, seq) do
+        {nil, waiters} ->
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: `nil`?? Could this hide a bug?

File: ex/loqui/lib/loqui/client.ex [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    end
+  end
+
+  defp to_host(host) when is_bitstring(host),
+    do: String.to_charlist(host)
+
+  defp to_host(host) when is_list(host),
+    do: host
+
+  defp next_sequence(%State{sequence: seq}=state) when seq >= @max_sequence do
```
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: I think the intention is for this to be `>` and not `>=` - not that it matters.

File: ex/loqui/lib/loqui/client.ex (hunk 7 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I'm being careful here, I don't want a server bug to take down the client.

File: ex/loqui/lib/loqui/protocol/parser.ex (hunk 1 above)
Comment (scohen) [inline ex/loqui/lib/loqui/protocol/parser.ex]: 🎓

File: ex/loqui/lib/loqui/client.ex (hunk 7 above)
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: Perhaps want to keep this but also log like has been added when a pong arrives? If you get such garbage from server is your client going to be able to assume its getting valid responses?

File: ex/loqui/lib/loqui/client.ex [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    with {:ok, sock}  <- :gen_tcp.connect(host, port, tcp_opts, connect_timeout),
+         {:ok, state} <- do_upgrade(%{state | sock: sock}),
+         state        <- make_active_once(state),
+         {:ok, state} <- do_loqui_connect(state) do
+
+      {:ok, state}
+    else
+      {:error, _} = error ->
+        Logger.error("Couldn't connect to #{host}:#{port} because #{inspect error}")
+        {:stop, error, state}
```
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Dumb question, but could we do this instead?
```elixir
with {:ok, sock} <- :gen_tcp.connect(host, port, tcp_opts, connect_timeout),
         state <- %{state | sock: sock},
         {:ok, state} <- do_upgrade(state),
 ...
```

File: ex/loqui/lib/loqui/client.ex [hunk 10]
```diff
@@ -0,0 +1,358 @@
+defmodule Loqui.Client do
+  defmodule State do
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: We normally don't create an additional module for the state.

File: ex/loqui/lib/loqui/client.ex [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+  end
+
+  alias Loqui.Protocol.{Codec, Codecs, Compressor, Compressors, Frames, Parser}
+  use Connection
+  use Loqui.{Opcodes, Types}
+  require Logger
+
+  @default_timeout 5000
+  @buffer_size 4096
+  @max_sequence round(:math.pow(2, 32) - 1)
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: 🤔  I wonder if this is a constant somewhere.

File: ex/loqui/lib/loqui/client.ex [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+    tcp_opts = opts
+      |> Keyword.get(:tcp_opts, [])
+      |> Keyword.put(:active, :false)
+      |> Keyword.put(:mode, :binary)
+      |> Keyword.put(:send_timeout_close, true)
+      |> Keyword.put_new(:buffer, @buffer_size)
+
+    {loqui_opts, _opts} = Keyword.pop(opts, :loqui_opts, [])
+
+    extract_option = fn(opt_name) ->
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: This seems like it should be a `defp` with `loqui_opts` passed in.

File: ex/loqui/lib/loqui/client.ex [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+    {waiter, new_waiters} = Map.pop(waiters, sequence)
+    decoded_data = codec.decode(payload)
+    Connection.reply(waiter, decoded_data)
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:go_away, _flags, close_code, payload_data}, %State{}=state) do
+    Process.send_after(self(), {:close_go_away, close_code, payload_data}, @go_away_timeout)
+    %State{state | sequence: :go_away}
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: It seems odd to call this `sequence` to me. Especially since we have a sequence number already.

File: ex/loqui/lib/loqui/client.ex [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+    go_away_packet = Frames.goaway(0, 0, "Closing")
+    :gen_tcp.send(sock, go_away_packet)
+
+    {:disconnect, {:close, from}, s}
+  end
+
+  def handle_call({:request, payload}, caller, %State{sock: sock, codec: codec}=state) do
+    {next_seq, state} = next_sequence(state)
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.request(0, next_seq, encoded_payload))
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Does it make sense to buffer requests and send them in batches? We got a big throughput boost on the server batching responses.

File: ex/loqui/lib/loqui/client.ex [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Do we ever sweep out these `waiters`? I can see this being a memory leak over time. Are we monitoring them?

File: ex/loqui/lib/loqui/client.ex [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+  end
+
+  def handle_cast(_, %State{sequence: :go_away}=state),
+    do: {:noreply, state}
+
+  def handle_cast({:push, payload}, %State{sock: sock, codec: codec}=state) do
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.push(0, encoded_payload))
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: We should probably batch here too.

File: ex/loqui/lib/loqui/client.ex [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+  def handle_info({:tcp, socket, data}, %State{sock: socket}=state) do
+    make_active_once(state)
+    with {:ok, parsed_packets, leftover_data} <- Parser.parse(state.buffer, data) do
+      state = handle_packets(parsed_packets, state)
+
+      {:noreply, %State{state | buffer: leftover_data}}
+    else
+
+      {:error, {:need_more_data, data}} ->
+        {:noreply, %State{buffer: data }}
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Extra space here after `data`

Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Are we dropping the old buffered data here? I don't see data + buffered data.

**Comment by fishcakez** [comment]
Else clauses of with can only use variables bound before the with
expression, so no. Likely need to separate connecting and handshaking.

File: ex/loqui/lib/loqui/client.ex [hunk 18]
```diff
@@ -0,0 +1,358 @@
...
+
+  def handle_info({:tcp_error, _, _}, _state) do
+    {:stop, :tcp_closed, nil}
+  end
+
+  def handle_info({:close_go_away, go_away_code, go_away_data}, %State{sequence: :go_away, waiters: waiters}=state) do
+    # tell the waiters that the remote end went away, then close.
+    err = {:error, {:remote_went_away, go_away_code, go_away_data}}
+    Enum.each(waiters, fn {_, waiter} ->
+      Connection.reply(waiter, err)
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Does `Connection.reply` wait for a response from the remote? If not, is there a more `push` like method? I don't think we need a response.

File: ex/loqui/lib/loqui/client.ex [hunk 19]
```diff
@@ -0,0 +1,358 @@
...
+
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+    new_state = state
+      |> schedule_ping
+      |> Map.put(:last_ping, {next_seq, :erlang.system_time(:milli_seconds)})
+
+    {:noreply, new_state}
+  end
+  def handle_info(:send_ping, %State{last_ping: {sequence, _ts}}=state) do
+    {:stop, {:error, {:pong_not_received, sequence}}, state}
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: 🤔  What happens if we have a pong in the mailbox but just haven't gotten to it?

File: ex/loqui/lib/loqui/client.ex [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+        {reply_to, waiters} ->
+          Connection.reply(reply_to, :pong)
+          waiters
+      end
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:hello_ack, _flags, ping_interval, data}, %State{}=state) do
+    [encoding, compression] = String.split(data, "|")
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: 🤔  I wonder if we can future proof this by not doing a hard match here. I'm thinking of when we update the server to start sending back more than 2 things. Would be nice to be able to update the server without breaking all the clients.

File: ex/loqui/lib/loqui/client.ex [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+    %State{state | sequence: :go_away}
+  end
+
+  defp handle_packet(packet, state) do
+    Logger.error("Received unknown packet, opcode #{inspect elem(packet, 0)} #{inspect packet}")
+    state
+  end
+
+  defp schedule_ping(%State{ping_interval: ping_interval}=state) do
+    Process.send_after(self(), :send_ping, ping_interval)
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: We should send pings faster than the ping interval, right? Is `ping_interval` the max before it will time out?

File: ex/loqui/lib/loqui/client.ex [hunk 22]
```diff
@@ -0,0 +1,358 @@
...
+    state
+  end
+
+  defp make_active_once(%State{sock: sock}=state) do
+    :ok = :inet.setopts(sock, [active: :once])
+
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: [nit] This line is not needed.

File: ex/loqui/lib/loqui/client.ex [hunk 23]
```diff
@@ -0,0 +1,358 @@
...
+    end
+  end
+
+  defp to_host(host) when is_bitstring(host),
+    do: String.to_charlist(host)
+
+  defp to_host(host) when is_list(host),
+    do: host
+
+  defp next_sequence(%State{sequence: seq}=state) when seq > @max_sequence do
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Should this be `>=`?

File: ex/loqui/lib/loqui/cowboy_protocol.ex [hunk 24]
```diff
@@ -120,22 +120,47 @@ defmodule Loqui.CowboyProtocol do
...
   defp handle_socket_data(state, data) do
-    case Protocol.handle_data(data) do
-      {:ok, request, extra} ->
-        case handle_request(request, state) do
-          {:ok, state} -> handle_socket_data(state, extra)
-          {:shutdown, reason} -> close(state, reason)
-        end
-      {:continue, extra} -> handler_loop(state, extra)
-      {:error, reason} -> goaway(state, reason)
+    with {:ok, decoded_packets, extra_data} <- Protocol.Parser.parse(data),
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/cowboy_protocol.ex]: This should be `decoded_requests` to match the rest of the code.

File: ex/loqui/lib/loqui/protocol/compressors/noop.ex [hunk 25]
```diff
@@ -0,0 +1,12 @@
+defmodule Loqui.Protocol.Compressors.NoOp do
+  @behaviour Loqui.Protocol.Compressor
+
+  def name,
+    do: " "
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/protocol/compressors/noop.ex]: Why an empty space?

Comment (jhowarth) [inline ex/loqui/lib/loqui/protocol/compressors/noop.ex]: Maybe name it `no-op`?

File: ex/loqui/lib/loqui/client.ex (hunk 12 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: it's only used here though.

File: ex/loqui/lib/loqui/client.ex (hunk 11 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Not that I could find. It's pretty arbitrary what the max seq is.

File: ex/loqui/lib/loqui/client.ex (hunk 14 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I can try, but we're getting 80k requests / sec now.

File: ex/loqui/lib/loqui/client.ex (hunk 15 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: We do, when when a response comes in, or a ping is returned. If the server dies, I suppose this could be a memory leak. Even if we monitor, there's no guarantee that the waiters will die; they too could be long lived.

File: ex/loqui/lib/loqui/client.ex (hunk 16 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: What performance goals should I shoot for?

File: ex/loqui/lib/loqui/client.ex (hunk 18 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Connection.reply is effectively GenServer.reaply; In this case, we've been given a go_away packet and have gone past our 1 second "Take no more requests and wait for responses to arrive" window. Since we haven't received anything yet, we send back an error.

File: ex/loqui/lib/loqui/client.ex (hunk 13 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: It's a marker, if the sequence is go_away, we're in an invalid state.

File: ex/loqui/lib/loqui/protocol/compressors/noop.ex (hunk 25 above)
Comment (scohen) [inline ex/loqui/lib/loqui/protocol/compressors/noop.ex]: It's an empty space because that's what the server sends. These names are the over-wire names, so naming it no-op would make it not work, since none of the servers call it no-op.

File: ex/loqui/lib/loqui/client.ex (hunk 23 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I don't think so, since max implies that it's allowed.

File: ex/loqui/lib/loqui/cowboy_protocol.ex (hunk 24 above)
Comment (scohen) [inline ex/loqui/lib/loqui/cowboy_protocol.ex]: ✅

File: ex/loqui/lib/loqui/client.ex (hunk 22 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: ✅

File: ex/loqui/lib/loqui/client.ex (hunk 21 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: The protocol didn't specify other than calling it the ping interval, in which case, I assumed it meant the interval at which we send pings. All the other clients and servers do something very similar, what's the advantage of sending this more frequently?

File: ex/loqui/lib/loqui/client.ex (hunk 20 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Doing so will have to increment the version of the protocol, no? If there's something added, clients will need to account for it and not drop it on the ground. At least, I hope that's the case.

File: ex/loqui/lib/loqui/client.ex (hunk 19 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I'm tying to come up with a circumstance where a healthy client can exhibit this behavior. If there's a pong in the mailbox and we get this message, that means it was delivered after the ping_interval, correct? If that's the case, it took too long.

File: ex/loqui/lib/loqui/client.ex (hunk 17 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: No, the data returned here is the old buffered data + the data read from the socket. Code in `packet.ex` takes care of the concatenation.

Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: ✅

File: ex/loqui/lib/loqui/client.ex (hunk 14 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I explored your suggestion, and while it might make sense on the server, where you do a lot of reading, the only packet here that involves a write in response to a read is a ping packet. Those happen sporadically enough that it didn't affect performance of the client at all to do the buffering.

File: ex/loqui/lib/loqui/client.ex (hunk 10 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I'd suggest revisiting that; especially on the server where you're doing the buffering. The add to buffer / flush dance could be expressed much more cleanly as:
```elixir
state 
 |> handle_packets
 |> State.flush_buffer
# in handle_packets 
Frame.reply(...)
|> State.add_to_buffer

# and State.flush
def flush_buffer(%State{socket: sock, buffered_frames: buffer}=state) do 
   :gen_tcp.write(sock, buffer)
   %{state | buffer: []}
end
```

File: ex/loqui/lib/loqui/client.ex (hunk 16 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Also, how should I decide to flush the buffer?

File: ex/loqui/lib/loqui/client.ex (hunk 10 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Check out the differences between 16c7cb0b3f5ed43369b159f94d9977428f46bfc1 and ab60dd8ff0ccff11a86e87949015553b26f2e283 for an example of how having a separate module for state cleans up code.

File: ex/loqui/lib/loqui/client.ex (hunk 16 above)
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: Probably send to `self()` after send, and accumulate a buffer until that message arrives, otherwise will delay sending. Hopefully everything can be accumulated without getting scheduled out too many times. You could limit the buffer size before the next send.

File: ex/loqui/lib/loqui/client.ex (hunk 19 above)
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: It's possible for the pong to be tcp recv buffer but if the server is responding instantly and this timeout has occurred its then the server is not going to feel responsive to the client - i.e.everything is queued for approximately the ping interval.

File: ex/loqui/lib/loqui/client.ex (hunk 21 above)
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: You may want to randomise the ping timer value slightly to avoid all the clients don't end up in sync on their pings.

File: ex/loqui/lib/loqui/client.ex (hunk 11 above)
Comment (fishcakez) [inline ex/loqui/lib/loqui/client.ex]: Over the wire this is encoded as a uint32 (unsigned 32 bit integer) so this is max sequence integer without overflowing. A larger number would overflow and a smaller sequence id would be sent back in a response.

File: ex/loqui/lib/loqui/client.ex (hunk 16 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Is that worth the extra complexity? We're currently getting 85K rps on a single client with 10 workers.

File: ex/loqui/lib/loqui/client.ex (hunk 11 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Ok, let me rephrase; As long as we don't overflow a 32 bit int, it's pretty arbitrary what the max sequence id is.

File: ex/loqui/lib/loqui/client.ex (hunk 19 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: Yes, the question is if we should account for that case. It feels like we have an overloaded server there.

File: ex/loqui/lib/loqui/client.ex (hunk 21 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: ✅

Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: ✅

File: ex/loqui/lib/loqui/client.ex (hunk 16 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I added a middle ground packet buffer, which does what @fishcakez suggested; puts a message at the end of your queue and buffers until that message is received.

Perf is roughly the same at 85-95k requests / sec, but it's more consistent.

**Comment by scohen** [comment]
@jhowarth re-review, please.

File: ex/loqui/lib/loqui/client.ex [hunk 26]
```diff
@@ -0,0 +1,420 @@
...
+      compressor: Compressors.NoOp,
+      codec: Codecs.Erlpack,
+      waiters: %{},
+      last_ping: nil
+
+    def add_packet(%State{packet_buffer: buffer}=state, packet) do
+      %State{state | packet_buffer: [packet | buffer]}
+    end
+
+    def push_waiter(%State{waiters: waiters}=state, sequence, waiter) do
```
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: [nit] This should be `put_waiter` to be consistent with a map.

File: ex/loqui/lib/loqui/client.ex (hunk 10 above)
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Seems ok, it's a bit more encapsulated. It sort of feels like we're just reimplementing map there.

File: ex/loqui/lib/loqui/client.ex (hunk 23 above)
Comment (jhowarth) [inline ex/loqui/lib/loqui/client.ex]: Doesn't that put the sequence at `max + 1` when it fails through to the function clause that adds 1 to it?

Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: It does, but then when `next_sequence` is called again, you'll go back to 1. If someone accessed the field directly, that'd be a problem though. Is that what you're getting at?

File: ex/loqui/lib/loqui/client.ex (hunk 26 above)
Comment (scohen) [inline ex/loqui/lib/loqui/client.ex]: I was going for the push/pop thing. Changed to add/remove. 

I disagree that we're recapitulating maps here, we're building a data structure using composition.

**Comment by scohen** [comment]
@jhgg review?

**Comment by scohen** [comment]
@jhgg; I'm going to merge. I have a subsequent PR that I'm going to post, and you can comment on that.

**Comment by jhgg** [comment]
@scohen retroactively looks good to me. didn't see anything obviously wrong :D
